psycopg_pool/psycopg_pool/pool_async.py

```diff
@@ -142,23 +141,41 @@ def __init__(
         if open:
             self._open()
 
+    if False:  # ASYNC
```

Comment:
No, it wouldn't be the same. I forgot to explain the rationale in the MR.

The Python compiler is able to recognise `if True` to always take the if-side branch and removes the if, and the whole else-side, from the code:

```
In [4]: def f():
   ...:     if True:
   ...:         return 10
   ...:     else:
   ...:         return 20
   ...: 

In [5]: dis.dis(f)
  2           0 NOP

  3           2 LOAD_CONST               2 (10)
              4 RETURN_VALUE
```

this doesn't happen using a generic expression:

```
In [6]: ASYNC=True

In [7]: def f():
   ...:     if ASYNC:
   ...:         return 10
   ...:     else:
   ...:         return 20
   ...: 

In [8]: dis.dis(f)
  2           0 LOAD_GLOBAL              0 (ASYNC)
              2 POP_JUMP_IF_FALSE        4 (to 8)

  3           4 LOAD_CONST               1 (10)
              6 RETURN_VALUE

  5     >>    8 LOAD_CONST               2 (20)
             10 RETURN_VALUE
```

Not only this works as an optimization, Mypy is also clever enough to apply the same consideration, so it doesn't complain if the else-side contains code that is invalid in the async context, but will become valid in the sync one (such as wrong return type, missing import because they are if'ed away too, etc).

However, I have just noticed that a constant string works too. Maybe it would be more readable to use `if "async"` or `if "ASYNC"` (any case) as a test?

Reply:
I've checked and, while Python is smart enough to prune branches under `if "async"` and `if not "async"`, MyPy is not so clever and raises error as if those branches were reachable.

We might raise a bug to MyPy to ask for more complete constant folding, but for the moment I think this solution is as good as it gets.

Reply:
I have found a related feature request in MyPy and added our use case: https://github.com/python/mypy/issues/12583